File: internal/convert/convert.go
// Package convert turns a LoRA file from one format into another via a managed
// Python venv (safetensors/mlx/numpy). The conversion logic lives in the
// embedded convert.py; this file manages the toolchain and orchestration.
//
// safetensors↔mlx are faithful; diffusers is a best-effort key remap; Draw
// Things read/write are experimental (proprietary float32-only handling).
package convert

import (
	"bufio"
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/keeandrews/loradex-cli/internal/config"
	"github.com/keeandrews/loradex-cli/internal/output"
)

//go:embed convert.py
var convertScript []byte

// Format is a LoRA serialization format.
type Format string

const (
	Safetensors Format = "safetensors"
	MLX         Format = "mlx"
	Diffusers   Format = "diffusers"
	DrawThings  Format = "drawthings"
)

// Targets are the formats convert can produce.
var Targets = []Format{Safetensors, MLX, Diffusers, DrawThings}

// Ext returns the output file extension for a format.
func (f Format) Ext() string {
	if f == DrawThings {
		return ".ckpt"
	}
	return ".safetensors" // mlx/diffusers/safetensors all use the safetensors container
}

// Result is the parsed outcome of a conversion.
type Result struct {
	OK       bool     `json:"ok"`
	Tensors  int      `json:"tensors"`
	Warnings []string `json:"warnings"`
	Quality  string   `json:"quality"` // faithful | experimental
	Error    string   `json:"error"`
}

// DetectFormat infers a file's source format from its name and magic bytes.
func DetectFormat(path string) (Format, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	magic := make([]byte, 16)
	_, _ = f.Read(magic)
	if bytes.HasPrefix(magic, []byte("SQLite format 3\x00")) {
		return DrawThings, nil
	}
	switch strings.ToLower(filepath.Ext(path)) {
	case ".safetensors":
		return Safetensors, nil
	case ".ckpt":
		return DrawThings, nil
	}
	// safetensors begins with an 8-byte little-endian header length; treat
	// anything else readable as safetensors and let the converter validate.
	return Safetensors, nil
}

// Convert runs convert.py for one (src→dst) conversion, writing outPath.
func Convert(ctx context.Context, src string, srcFmt, dstFmt Format, outPath string, p *output.Printer) (Result, error) {
	python, err := EnsureTools(ctx, p, dstFmt == MLX)
	if err != nil {
		return Result{}, err
	}
	script, err := writeScript()
	if err != nil {
		return Result{}, err
	}
	if err := os.MkdirAll(filepath.Dir(outPath), 0o755); err != nil {
		return Result{}, err
	}
	cmd := exec.CommandContext(ctx, python, script, src, string(srcFmt), string(dstFmt), outPath)
	cmd.Env = cleanEnv()
	var stdout bytes.Buffer
	cmd.Stdout = &stdout
	stderr, _ := cmd.StderrPipe()
	if err := cmd.Start(); err != nil {
		return Result{}, output.Errorf(output.ExitError, "convert_failed", "", "could not start converter: %v", err)
	}
	go func() {
		sc := bufio.NewScanner(stderr)
		for sc.Scan() {
			if p.ProgressEnabled() {
				fmt.Fprintf(p.Err, "  %s\n", sc.Text())
			}
		}
	}()
	waitErr := cmd.Wait()

	var res Result
	if line := lastJSONLine(stdout.Bytes()); line != nil {
		_ = json.Unmarshal(line, &res)
	}
	if !res.OK {
		msg := res.Error
		if msg == "" && waitErr != nil {
			msg = waitErr.Error()
		}
		return res, output.Errorf(output.ExitError, "convert_failed",
			"check the converter output above", "conversion failed: %s", orStr(msg, "unknown error"))
	}
	return res, nil
}

// EnsureTools creates/returns the managed conversion venv python, installing
// safetensors+numpy (and mlx when needed). Reuses the venv across runs.
func EnsureTools(ctx context.Context, p *output.Printer, needMLX bool) (string, error) {
	base, err := config.Dir()
	if err != nil {
		return "", err
	}
	toolsDir := filepath.Join(base, "tools")
	venv := filepath.Join(toolsDir, "venv")
	venvPy := filepath.Join(venv, "bin", "python")
	if runtime.GOOS == "windows" {
		venvPy = filepath.Join(venv, "Scripts", "python.exe")
	}

	if !isFile(venvPy) {
		py3, err := findPython3()
		if err != nil {
			return "", output.Errorf(output.ExitValidation, "python_missing",
				"install Python 3.10+ (python3)", "python3 is required for format conversion")
		}
		p.Info("setting up the conversion toolchain (one-time)…")
		if err := run(ctx, p, py3, "-m", "venv", venv); err != nil {
			return "", err
		}
	}
	if !importsOK(ctx, venvPy, "safetensors", "numpy") {
		p.Info("installing conversion dependencies (safetensors, numpy)…")
		_ = run(ctx, p, venvPy, "-m", "pip", "install", "--upgrade", "pip")
		if err := run(ctx, p, venvPy, "-m", "pip", "install", "safetensors", "numpy"); err != nil {
			return "", output.Errorf(output.ExitError, "deps_failed", "", "could not install conversion deps: %v", err)
		}
	}
	if needMLX && !importsOK(ctx, venvPy, "mlx.core") {
		p.Info("installing MLX (Apple Silicon)…")
		if err := run(ctx, p, venvPy, "-m", "pip", "install", "mlx"); err != nil {
			return "", output.Errorf(output.ExitValidation, "mlx_unavailable",
				"MLX requires Apple Silicon; pick another target format", "could not install MLX: %v", err)
		}
	}
	return venvPy, nil
}

func writeScript() (string, error) {
	base, err := config.Dir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(base, "tools")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	path := filepath.Join(dir, "convert.py")
	if err := os.WriteFile(path, convertScript, 0o644); err != nil {
		return "", err
	}
	return path, nil
}

func importsOK(ctx context.Context, python string, mods ...string) bool {
	args := []string{"-c", "import " + strings.Join(mods, ", ")}
	cmd := exec.CommandContext(ctx, python, args...)
	cmd.Env = cleanEnv()
	return cmd.Run() == nil
}

func findPython3() (string, error) {
	for _, n := range []string{"python3", "python"} {
		if p, err := exec.LookPath(n); err == nil {
			return p, nil
		}
	}
	return "", fmt.Errorf("python3 not found")
}

func run(ctx context.Context, p *output.Printer, name string, args ...string) error {
	cmd := exec.CommandContext(ctx, name, args...)
	cmd.Env = cleanEnv()
	stdout, _ := cmd.StdoutPipe()
	stderr, _ := cmd.StderrPipe()
	if err := cmd.Start(); err != nil {
		return output.Errorf(output.ExitError, "exec_failed", "", "could not start %s: %v", filepath.Base(name), err)
	}
	relay := func(r io.Reader) {
		sc := bufio.NewScanner(r)
		sc.Buffer(make([]byte, 64*1024), 1024*1024)
		for sc.Scan() {
			if p.ProgressEnabled() {
				fmt.Fprintf(p.Err, "  %s\n", sc.Text())
			}
		}
	}
	done := make(chan struct{}, 2)
	go func() { relay(stdout); done <- struct{}{} }()
	go func() { relay(stderr); done <- struct{}{} }()
	<-done
	<-done
	if err := cmd.Wait(); err != nil {
		return output.Errorf(output.ExitError, "exec_failed", "", "%s failed: %v", filepath.Base(name), err)
	}
	return nil
}

func lastJSONLine(b []byte) []byte {
	var last []byte
	for _, ln := range bytes.Split(b, []byte("\n")) {
		t := bytes.TrimSpace(ln)
		if len(t) > 0 && t[0] == '{' {
			last = t
		}
	}
	return last
}

func cleanEnv() []string {
	src := os.Environ()
	out := make([]string, 0, len(src))
	for _, kv := range src {
		if strings.HasPrefix(kv, "LORADEX_") {
			continue
		}
		out = append(out, kv)
	}
	return out
}

func isFile(p string) bool {
	fi, err := os.Stat(p)
	return err == nil && !fi.IsDir()
}

func orStr(s, def string) string {
	if s == "" {
		return def
	}
	return s
}

File: internal/convert/convert.py
#!/usr/bin/env python3
"""loradex format converter.

Reads a LoRA from one format into a common dict[str, np.ndarray] (+ metadata),
then writes it to a target format. Invoked by the loradex CLI as:

    python convert.py <src_path> <src_fmt> <dst_fmt> <out_path>

Formats: safetensors | mlx | diffusers | drawthings
Emits a single JSON line on stdout: {"ok", "tensors", "warnings", "quality"}.

Reliability: safetensors/mlx are faithful (same tensors, MLX-validated container).
diffusers is a best-effort key remap. drawthings read/write are EXPERIMENTAL
(Draw Things uses a proprietary SQLite/NNC layout; only float32 LoRAs are handled).
"""
import json
import sqlite3
import struct
import sys

import numpy as np

WARNINGS = []


def warn(msg):
    WARNINGS.append(msg)


# ---------- readers ----------

def read_safetensors(path):
    from safetensors import safe_open
    tensors, meta = {}, {}
    with safe_open(path, framework="numpy") as f:
        meta = f.metadata() or {}
        for k in f.keys():
            tensors[k] = f.get_tensor(k)
    return tensors, meta


def _parse_dim(blob):
    if not blob or len(blob) < 8:
        return []
    n = len(blob) // 4
    vals = struct.unpack("<%di" % n, blob[: n * 4])
    dims = []
    for v in vals[1:]:  # vals[0] is a format/type header
        if v <= 0:
            break
        dims.append(v)
    return dims


def read_drawthings(path):
    """EXPERIMENTAL: extract float32 tensors from a Draw Things .ckpt (SQLite)."""
    warn("Draw Things read is experimental — only float32 LoRAs are supported")
    con = sqlite3.connect("file:%s?mode=ro" % path, uri=True)
    tensors = {}
    skipped = 0
    for name, dim, data in con.execute("SELECT name, dim, data FROM tensors"):
        shape = _parse_dim(dim)
        count = int(np.prod(shape)) if shape else 0
        if count <= 0 or not data or len(data) < count * 4:
            skipped += 1
            continue
        raw = bytes(data)[len(data) - count * 4:]  # tail bytes = raw f32 payload
        try:
            arr = np.frombuffer(raw, dtype="<f4").reshape(shape).astype(np.float32)
        except Exception:
            skipped += 1
            continue
        tensors[name] = arr
    con.close()
    if skipped:
        warn("skipped %d tensor(s) that weren't decodable as float32" % skipped)
    if not tensors:
        raise RuntimeError("no float32 tensors could be decoded from the Draw Things file")
    return tensors, {"source": "drawthings"}


# ---------- writers ----------

def write_safetensors(tensors, meta, path):
    from safetensors.numpy import save_file
    md = {k: str(v) for k, v in (meta or {}).items()}
    save_file({k: np.ascontiguousarray(v) for k, v in tensors.items()}, path, metadata=md or None)


def write_mlx(tensors, meta, path):
    """Re-container as an MLX-validated safetensors (every tensor round-tripped
    through mx.array, so MLX is guaranteed to load it)."""
    try:
        import mlx.core as mx
    except Exception as e:
        raise RuntimeError("MLX is not available on this machine (%s)" % e)
    arrs = {k: mx.array(np.ascontiguousarray(v)) for k, v in tensors.items()}
    md = {k: str(v) for k, v in (meta or {}).items()}
    mx.save_safetensors(path, arrs, metadata=md or {})


def _to_diffusers_key(k):
    # Best-effort kohya -> diffusers/PEFT renaming.
    k = k.replace("lora_unet_", "").replace("lora_te_", "te_")
    k = k.replace(".lora_down.weight", ".lora_A.weight")
    k = k.replace(".lora_up.weight", ".lora_B.weight")
    k = k.replace("_lora_down", ".lora_A").replace("_lora_up", ".lora_B")
    return k


def write_diffusers(tensors, meta, path):
    warn("diffusers conversion remaps keys best-effort; verify against your loader")
    remapped, alpha = {}, 0
    for k, v in tensors.items():
        if k.endswith(".alpha"):
            alpha += 1
            continue  # diffusers/PEFT derives scaling from rank, not an alpha tensor
        remapped[_to_diffusers_key(k)] = v
    if alpha:
        warn("dropped %d .alpha tensor(s) (not used by diffusers/PEFT)" % alpha)
    write_safetensors(remapped, {"format": "diffusers"}, path)


def write_drawthings(tensors, meta, path):
    """EXPERIMENTAL: write float32 tensors into a Draw Things-style SQLite. The
    resulting file may not load in Draw Things (its loader expects a specific NNC
    layout + a loratrainingconfiguration record). Provided for round-tripping."""
    warn("Draw Things write is EXPERIMENTAL and may not load in Draw Things itself")
    con = sqlite3.connect(path)
    con.execute(
        "CREATE TABLE tensors (name TEXT, type INTEGER, format INTEGER, "
        "datatype INTEGER, dim BLOB, data BLOB, PRIMARY KEY (name))"
    )
    for name, v in tensors.items():
        v = np.ascontiguousarray(v.astype(np.float32))
        dims = list(v.shape)
        dim_blob = struct.pack("<i", 0x00000C00) + b"".join(struct.pack("<i", d) for d in dims)
        dim_blob += struct.pack("<i", 0) * (8 - len(dims))
        con.execute(
            "INSERT INTO tensors (name, type, format, datatype, dim, data) VALUES (?,?,?,?,?,?)",
            (name, 0, 0, 1, dim_blob, v.tobytes()),
        )
    con.commit()
    con.close()


READERS = {"safetensors": read_safetensors, "drawthings": read_drawthings}
WRITERS = {
    "safetensors": write_safetensors,
    "mlx": write_mlx,
    "diffusers": write_diffusers,
    "drawthings": write_drawthings,
}


def main():
    if len(sys.argv) != 5:
        print(json.dumps({"ok": False, "error": "usage: convert.py <src> <src_fmt> <dst_fmt> <out>"}))
        return 2
    src, src_fmt, dst_fmt, out = sys.argv[1:5]
    if src_fmt not in READERS:
        print(json.dumps({"ok": False, "error": "unsupported source format %r" % src_fmt}))
        return 2
    if dst_fmt not in WRITERS:
        print(json.dumps({"ok": False, "error": "unsupported target format %r" % dst_fmt}))
        return 2
    try:
        tensors, meta = READERS[src_fmt](src)
        WRITERS[dst_fmt](tensors, meta, out)
    except Exception as e:  # noqa: BLE001
        print(json.dumps({"ok": False, "error": str(e), "warnings": WARNINGS}))
        return 1
    if src_fmt == "drawthings" or dst_fmt == "drawthings":
        quality = "experimental"
    elif dst_fmt == "diffusers":
        quality = "best-effort"
    else:
        quality = "faithful"
    print(json.dumps({
        "ok": True,
        "tensors": len(tensors),
        "warnings": WARNINGS,
        "quality": quality,
    }))
    return 0


if __name__ == "__main__":
    sys.exit(main())
